File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/ui/viewmodel/MeteoriteViewModel.java
package com.antonio.samir.meteoritelandingsspots.ui.viewmodel;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.content.Context;

import com.antonio.samir.meteoritelandingsspots.model.Meteorite;
import com.antonio.samir.meteoritelandingsspots.presenter.MeteoriteListPresenter;

import java.util.List;


public class MeteoriteViewModel extends AndroidViewModel {

    private MeteoriteListPresenter presenter;
    private LiveData<List<Meteorite>> meteorites = null;

    public MeteoriteViewModel(final Application application) {
        super(application);
    }

    public MeteoriteListPresenter getPresenter() {
        if (presenter == null) {
            final Context context = getApplication().getApplicationContext();
            presenter = new MeteoriteListPresenter(context);
        }
        return presenter;
    }

    public LiveData<List<Meteorite>> getMeteorites() {
        if (meteorites == null) {
            meteorites = presenter.getMeteorites();
        }
        return meteorites;
    }
}

File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/ui/viewmodel/MeteoriteViewModel.kt
package com.antonio.samir.meteoritelandingsspots.ui.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import com.antonio.samir.meteoritelandingsspots.model.Meteorite
import com.antonio.samir.meteoritelandingsspots.presenter.MeteoriteListPresenter


class MeteoriteViewModel(application: Application) : AndroidViewModel(application) {

    private var presenter: MeteoriteListPresenter? = null
    private var meteorites: LiveData<List<Meteorite>>? = null

    fun getPresenter(): MeteoriteListPresenter {
        if (presenter == null) {
            val context = getApplication<Application>().applicationContext
            presenter = MeteoriteListPresenter(context)
        }
        return presenter as MeteoriteListPresenter
    }

    fun getMeteorites(): LiveData<List<Meteorite>>? {
        if (meteorites == null) {
            meteorites = presenter!!.meteorites
        }
        return meteorites
    }
}
